common/archiver/s3store/historyArchiver.go

```diff
@@ -0,0 +1,333 @@
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/s3"
+	"github.com/aws/aws-sdk-go/service/s3/s3iface"
+
+	"github.com/uber/cadence/.gen/go/shared"
+	"github.com/uber/cadence/common"
+	"github.com/uber/cadence/common/archiver"
+	"github.com/uber/cadence/common/backoff"
+	"github.com/uber/cadence/common/log/tag"
+	"github.com/uber/cadence/common/service/config"
+)
+
+const (
+	// URIScheme is the scheme for the s3 implementation
+	URIScheme               = "s3"
+	errBucketNotExists      = "requested bucket does not exist"
+	errEncodeHistory        = "failed to encode history batches"
+	errWriteKey             = "failed to write history to s3"
+	defaultBlobstoreTimeout = 10 * time.Second
+	targetHistoryBlobSize   = 2 * 1024 * 1024 // 2MB
+)
+
+var (
+	errNoBucketSpecified = errors.New("no bucket specified")
+	errEmptyAwsRegion    = errors.New("empty aws region")
+)
+
+type (
+	historyArchiver struct {
+		container *archiver.HistoryBootstrapContainer
+		s3cli     s3iface.S3API
+		// only set in test code
+		historyIterator archiver.HistoryIterator
+		config          *config.S3Archiver
+	}
+
+	getHistoryToken struct {
+		CloseFailoverVersion int64
+		NextBatchIdx         int
+	}
+)
+
+// NewHistoryArchiver creates a new archiver.HistoryArchiver based on s3
+func NewHistoryArchiver(
+	container *archiver.HistoryBootstrapContainer,
+	config *config.S3Archiver,
+) (archiver.HistoryArchiver, error) {
+	return newHistoryArchiver(container, config, nil)
+}
+
+func newHistoryArchiver(
+	container *archiver.HistoryBootstrapContainer,
+	config *config.S3Archiver,
+	historyIterator archiver.HistoryIterator,
+) (*historyArchiver, error) {
+	if len(config.Region) == 0 {
+		return nil, errEmptyAwsRegion
+	}
+	s3Config := &aws.Config{
+		Endpoint:         config.Endpoint,
+		Region:           aws.String(config.Region),
+		S3ForcePathStyle: aws.Bool(config.S3ForcePathStyle),
+		MaxRetries:       aws.Int(0),
+	}
+	sess, err := session.NewSession(s3Config)
+	if err != nil {
+		return nil, err
+	}
+
+	return &historyArchiver{
+		container:       container,
+		s3cli:           s3.New(sess),
+		historyIterator: historyIterator,
+	}, nil
+}
+func (h *historyArchiver) Archive(
+	ctx context.Context,
+	URI archiver.URI,
+	request *archiver.ArchiveHistoryRequest,
+	opts ...archiver.ArchiveOption,
+) (err error) {
+	ctx, cancel := ensureContextTimeout(ctx)
+	defer cancel()
+	featureCatalog := archiver.GetFeatureCatalog(opts...)
+	defer func() {
+		if err != nil && !common.IsPersistenceTransientError(err) && !isRetryableError(err) && featureCatalog.NonRetriableError != nil {
+			err = featureCatalog.NonRetriableError()
+		}
+	}()
+
+	logger := archiver.TagLoggerWithArchiveHistoryRequestAndURI(h.container.Logger, request, URI.String())
+
+	if err := h.ValidateURI(URI); err != nil {
+		logger.Error(archiver.ArchiveNonRetriableErrorMsg, tag.ArchivalArchiveFailReason(archiver.ErrReasonInvalidURI), tag.Error(err))
+		return err
```

Comment:
Should be an non-retryable error

Reply:
I wrapped it in a BadRequestError, not sure if this is the correct pattern?

Reply:
We typically use BadRequestError when the input comes from the user. How about return some special error and check those errors in the `isRetryableError` function? 

I think there's a predefined error for invalid URI in the `archiver` package (constants.go). You can add another one for invalid archive request in that file.

Reply:
The error returned by ValidateURI would be archiver.ErrURISchemeMismatch or errNoBucketSpecified. isRetryableError would return false for these errors.  So I am not relying on that fact. So how the code looked originally would not have been retried either afaik. Maybe we can have a chat about this to clarify how I should do this properly :)

Reply:
Sorry, My bad. Please ignore all my comments regarding this. I had some misunderstanding for the `isRetryableError` function. Your first version (return err directly) is correct. My apologies.